lib/THC/THCGeneral.c

```diff
@@ -42,21 +101,153 @@ void THCudaInit(THCState* state)
 
             continue;
           }
+
           THCudaCheck(err);
         }
       }
     }
   }
-  THCudaCheck(cudaSetDevice(device));
+
+  /* Restore previous device before continuing */
+  THCudaCheck(cudaSetDevice(prevDev));
 }
 
-void THCudaShutdown(THCState* state)
+int THCState_getNumDevices(THCState *state)
 {
-  THCRandom_shutdown(state);
-  free(state->blasState);
-  free(state->rngState);
-  free(state->deviceProperties);
-  THCudaBlas_shutdown(state);
+  return state->numDevices;
+}
+
+void THCState_reserveStreams(THCState* state, int numStreams)
+{
+  if (numStreams <= state->numUserStreams)
+  {
+    return;
+  }
+
+  int prevDev = -1;
+  THCudaCheck(cudaGetDevice(&prevDev));
+
+  /* Otherwise, we have to allocate a new set of streams */
+  cudaStream_t** newStreams =
+    (cudaStream_t**) malloc(state->numDevices * sizeof(cudaStream_t*));
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    THCudaCheck(cudaSetDevice(dev));
+
+    /* +1 for the default stream as well */
+    newStreams[dev] =
+      (cudaStream_t*) malloc((numStreams + 1) * sizeof(cudaStream_t));
+
+    /* Copy over old streams
+       (0 is default stream, 1 ... numUserStreams are rest) */
+    for (int stream = 0; stream <= state->numUserStreams; ++stream) {
+      newStreams[dev][stream] = state->streamsPerDevice[dev][stream];
+    }
+
+    /* Allocate new streams */
+    for (int stream = state->numUserStreams + 1; stream <= numStreams; ++stream) {
+      newStreams[dev][stream] = NULL;
+      THCudaCheck(cudaStreamCreate(&newStreams[dev][stream]));
+    }
+  }
+
+  cudaStream_t** oldStreams = state->streamsPerDevice;
+  state->streamsPerDevice = newStreams;
+  state->numUserStreams = numStreams;
+
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    free(oldStreams[dev]);
+  }
+  free(oldStreams);
+
+  THCudaCheck(cudaSetDevice(prevDev));
+}
+
+int THCState_getNumStreams(THCState* state)
+{
+  return state->numUserStreams;
+}
+
+void THCState_resetStreams(THCState* state, int device)
+{
+  if (state->currentStream !=
+      state->streamsPerDevice[device][state->currentPerDeviceStream]) {
+    THError("Unexpected stream state");
+  }
+
+  /* Reallocate all streams for the current device; the 0 stream
+     doesn't need updating */
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    for (int stream = 1; stream <= state->numUserStreams; ++stream) {
+      THCudaCheck(cudaStreamCreate(&state->streamsPerDevice[dev][stream]));
```

Comment:
No, cudaDeviceReset() destroys all state associated with that device; the device no longer knows about previous streams.

There were a couple of different bugs with the device reset in cutorch; one was that pointers to various states weren't updated, and the second is that p2p device access was also destroyed and needs to be restored.

It's a really dangerous call, I argue that device reset should be removed from cutorch entirely, especially since we have no way of resetting/cordoning off all memory allocated for that device previously.

e.g.,
```
void* p = NULL;
cudaMalloc(&p, 1000);
cudaDeviceReset();
cudaError_t err = cudaFree(p);
assert(err == cudaSuccess);
```

will fail, just as:

```
cudaStream_t stream;
cudaStreamCreate(&stream);
cudaDeviceReset();
cudaError_t err = cudaStreamDestroy(stream);
assert(err == cudaSuccess);
```

will fail.